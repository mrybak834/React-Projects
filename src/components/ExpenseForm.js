import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();

export default class ExpenseForm extends React.Component {

    constructor(props){
        super(props);

        this.state = {
            description: '',
            note: '',
            ...props.expense,
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            timestamp: props.expense ? moment(props.expense.timestamp) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;

        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;

        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;

        // Accept no amount or a valid value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onCalendarDateChange = (timestamp) => {
        timestamp && this.setState(() => ({ timestamp }));
    };

    onCalendarFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused}));
    };

    onSubmit = (e) => {
        e.preventDefault();

        // Handle input errors
        this.setState(() => ({ error: '' }));
        if (!this.state.description.trim()){
            return this.setState(() => ({ error: 'Please enter a description!'}));
        }
        if (!this.state.amount){
            return this.setState(() => ({ error: 'Please enter an amount!'}));
        }

        this.props.onSubmit({
            description: this.state.description,
            note: this.state.note,
            amount: parseFloat(this.state.amount, 10) * 100,
            timestamp: this.state.timestamp.valueOf()
        });
    };

    render() {
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    <input 
                        id="inputDescription"
                        type="text" 
                        placeholder="Description" 
                        autoFocus 
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        id="inputAmount"
                        type="text" 
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.timestamp}
                        onDateChange={this.onCalendarDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onCalendarFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={(day) => false}
                    />
                    <textarea 
                        placeholder="Add a note for your expense (Optional)" 
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    />
                    <button >Add Expense</button>
                </form>
                {this.state.error && <p>{this.state.error}</p>}
            </div>
        )
    }
}