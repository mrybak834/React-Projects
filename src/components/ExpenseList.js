import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../redux/selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No Expenses</p>
            ) : (
                props.expenses.map((expense, index) => 
                    <ExpenseListItem 
                        key={index}
                        {...expense}/>
                )
            )
        }
    </div>
);


const mapStateToProps = (state) => ({
        expenses: selectExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);