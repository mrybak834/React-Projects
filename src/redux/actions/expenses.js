import uuid from 'uuid';


export const addExpense = ( 
    {
        description = '', 
        note = '', 
        amount = 0, 
        timestamp = 0
    } = {} 
) => (
    {
        type: 'ADD_EXPENSE',
        expense: {
            id: uuid(),
            description,
            note,
            amount,
            timestamp
        }
    }
);

export const removeExpense = (id = undefined) => (
    {
        type: 'REMOVE_EXPENSE',
        id
    }
);

export const editExpense = (id = undefined, updates = {}) => (
    {
        type: 'EDIT_EXPENSE',
        id,
        updates
    }
);