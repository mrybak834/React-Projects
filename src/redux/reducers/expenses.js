const expensesReducerDefaultState = [];

export default ( state = expensesReducerDefaultState, action ) => {
    switch (action.type){
        case 'ADD_EXPENSE':{
            if (!action.expense){
                return state;
            }
            return state.concat(action.expense);
            // return [...state, action.expense];   // Slower than concat
        }
        case 'REMOVE_EXPENSE':{
            // No id defined, don't remove anything
            if (!action.id){
                return state;
            }

            return state.filter(({id}) => id !== action.id);
        }
        case 'EDIT_EXPENSE':{
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        // Overrides any of the expense properties above
                        ...action.updates
                    }
                }

                return expense;
            });
        }
        default:{
            return state;
        }
    }
};