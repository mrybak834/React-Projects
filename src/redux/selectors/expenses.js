import moment from 'moment';


export default (expenses, { text, sortBy, startDate, endDate }) => {

    return expenses.filter( (expense) => {
        // // Filter text
        // if ( text && ((expense.description !== text) && (expense.note !== text)) ){
        //     return false;
        // }
        // // Filter Dates
        // if ( (startDate && (expense.timestamp < startDate)) || (endDate && (expense.timestamp > endDate))){
        //     return false;
        // }

        const timestampMoment = moment(expense.timestamp);

        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase());
        const startDateMatch = startDate ? startDate.isSameOrBefore(timestampMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(timestampMoment, 'day') : true;
        return textMatch && startDateMatch && endDateMatch;
    }).sort( (a,b) => {
        if (sortBy === 'date'){
            // -1 = a comes first
            return a.timestamp > b.timestamp ? -1 : 1;
        }
        // amount
        else {
            return a.amount > b.amount ? -1 : 1;
        }
    });
};