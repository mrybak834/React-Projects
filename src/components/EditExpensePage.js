import React from 'react';
import ExpenseForm from '../components/ExpenseForm';
import { connect } from 'react-redux';
import { editExpense } from '../redux/actions/expenses';
import { removeExpense } from '../redux/actions/expenses';

export class EditExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);

        this.props.history.push('/');
    };

    onClick = () => {
        this.props.removeExpense(this.props.expense.id);
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <p>Edit Expense: {this.props.expense.description}</p>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button 
                    onClick={this.onClick}
                >Remove</button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id,expense)),
    removeExpense: (id) => dispatch(removeExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);