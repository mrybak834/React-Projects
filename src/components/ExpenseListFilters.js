import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, setSortByFilter, setStartDateFilter, setEndDateFilter } from '../redux/actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    
    state = {
        calendarFocused: null
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSelectChange = (e) => {
        this.props.setSortByFilter(e.target.value);
    };

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDateFilter(startDate);
        this.props.setEndDateFilter(endDate);
    };

    render () {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}
                />
                
                <select 
                    value={this.props.filters.sortBy}
                    onChange={this.onSelectChange}
                >
                    <option value="date">Date</option>
                    <option value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    startDateId="start_date"
                    endDate={this.props.filters.endDate}
                    endDateId="end_date"
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={2}
                    isOutsideRange={() => false}
                />
        
            </div>
        );
    }
}


const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};

const mapDispatchToProps = (dispatch, props) => ({
    setTextFilter: value => dispatch(setTextFilter(value)),
    setSortByFilter: value => dispatch(setSortByFilter(value)),
    setStartDateFilter: startDate => dispatch(setStartDateFilter(startDate)),
    setEndDateFilter: endDate => dispatch(setEndDateFilter(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);