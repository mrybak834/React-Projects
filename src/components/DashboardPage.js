import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';


const DashboardPage = () => (
    <div>
        <p>Dashboard Page</p>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default DashboardPage;