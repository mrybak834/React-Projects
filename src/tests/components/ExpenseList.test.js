/**
 * Export the unconnected component in order to use it for testing,
 * since we are adding dummy data
 */
import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';


test('Components: ExpenseList: Render ExpenseList (populated)', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses}/>);

    expect(wrapper).toMatchSnapshot();
});

test('Components: ExpenseList: Render ExpenseList (empty)', () => {
    const wrapper = shallow(<ExpenseList expenses={[]}/>);

    expect(wrapper).toMatchSnapshot();
});