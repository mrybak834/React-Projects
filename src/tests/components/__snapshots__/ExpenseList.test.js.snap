// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: ExpenseList: Render ExpenseList (empty) 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;

exports[`Components: ExpenseList: Render ExpenseList (populated) 1`] = `
<div>
  <ExpenseListItem
    amount={1}
    description="test"
    id="1"
    key="0"
    note="test"
    timestamp={864000000}
  />
  <ExpenseListItem
    amount={60000}
    description="Rent"
    id="2"
    key="1"
    note=""
    timestamp={432000000}
  />
  <ExpenseListItem
    amount={99000}
    description="Car"
    id="3"
    key="2"
    note="Car payment"
    timestamp={0}
  />
  <ExpenseListItem
    amount={4500}
    description="Credit Card"
    id="4"
    key="3"
    note="Mastercard"
    timestamp={-432000000}
  />
</div>
`;
